### ISSUE TITLE
Application Instance in access logs?

### ISSUE THREAD

**Issue by altenhof** [cloudfoundry/gorouter#136]
Hi there,

I was wondering whether there is a way to determine from an access log message to which application instance the request was routed. I assume the "source_instance" field that shows up in a firehose LogMessage is referring to the gorouter instance, not the application instance, right?

Thanks
Michael Altenhofen

**Comment by crhino** [comment]
@altenhof The `source_instance` field is referring the gorouter instance in this case, correct.

The access logs do not currently have any information around what application instance the request was routed to. What type of information would you want to see? What use cases do you have?

**Comment by altenhof** [comment]
@crhino

we at SAP try to use the logging infrastructure to add value to our developers when it comes to application performance issues. For that, we run an ELK cluster with fairly elaborated Logstash rules and quite a bunch of dedicated Kibana dashboards. In that setup, the logs stored in ES do have quite a number of fields that allow app developers to slice and dice their data. 

Recently, teams where trying to hunt down a performance issue (significant delay) between the response time reported by the RTR instance and the app. instance. They actually tried to correlate which RTR instance did talk to which app. instance in that particular case, but "failed" using the "component_instance" (as we call it) which we falsely derived in our rule set from the "source _instance".  

We're not there yet, but one could also imagine a graphical visualization of the request flow across the various components and having a simple way to determine where to draw edges would be helpful. 

So, in summary, it would increase usability for our setup, if a RTR log would also include the "app instance index", but I understand that this would lead to a further "complication", as you would have to extend the log message payload to include that additional fields. AS we would have to adapt our Logstash rules once the change had been implemented.

On the other hand, adding that info "later" in our pipeline seems like a major effort which we haven't fully addressed yet.

HTH

Michael Altenhofen

**Comment by crhino** [comment]
@altenhof 

Thanks for giving us more information around your use case. I can definitely see how having the ability to see which specific instance of an application a request was sent to would be helpful. I could potentially see adding a `instance_id` field to the access logs. This id would be the GUID associated with the running application container. The problem with that solution is that currently the `instance_id` of an application instance is not available from any of CF's public APIs  (you can see it as an environment variable in the container).

Would adding this `instance_id` field solve your use case, given that you would have to do some extra work to correlate that id with an application instance index and/or ip:port address?

@crhino && @jzlwang

**Comment by altenhof** [comment]
@crhino and @jzlwang 

Thanks for the update. 

Well, an `instance_id` GUID as part of the access log would certainly close the gap from a technical point-of-view. As we're currently want to avoid any time-consuming post processing of logs on top of what we already (have to) do, that would most likely lead to a solution that would be less user-friendly. We need to further investigate what that would require on our end.

Let's assume for a second that you would add such a field to a access log. Where would it show up? The reason why I'm asking is that the Logstash rules for parsing such RTR logs are already quite elaborated and having to deal with a additional optional fields that may show up in arbitrary order may cause some trouble. Have you ever thought about adding "custom fields" to the log line encoded as an (embedded) JSON object?

Best

@altenhof

**Comment by shalako** [comment]
We recently added the app instance index to the access log message: https://www.pivotaltracker.com/story/show/127325899

When a developer runs `cf logs myapp` they will see that the log message for the RTR includes the same identifier used in response to `cf app myapp`.

Does this fulfill your use case?

**Comment by altenhof** [comment]
Hi Shannon,

yes, it does. Thanks for adding this feature. When will that be released?
We need some heads-up to see whether our Logstash rules will still work. Is there a sample log line stored somewhere that we could use for unit testing?

Thanks

Michael

Von: Shannon Coen <notifications@github.com<mailto:notifications@github.com>>
Antworten an: cloudfoundry/gorouter <reply@reply.github.com<mailto:reply@reply.github.com>>
Datum: Mittwoch, 17. August 2016 um 03:35
An: cloudfoundry/gorouter <gorouter@noreply.github.com<mailto:gorouter@noreply.github.com>>
Cc: "Altenhofen, Michael" <michael.altenhofen@sap.com<mailto:michael.altenhofen@sap.com>>, Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Betreff: Re: [cloudfoundry/gorouter] Application Instance in access logs? (#136)


We recently added the app instance index to the access log message: https://www.pivotaltracker.com/story/show/127325899

When a developer runs cf logs myapp they will see that the log message for the RTR includes the same identifier used in response to cf app myapp.

Does this fulfill your use case?

**Comment by shalako** [comment]
Hello Michael,

Here's what a developer would see in response to `cf logs dora`:
```
2016-08-17T09:56:36.10-0700 [RTR/0]      OUT dora-test-scoen.superman.cf-app.com - [17/08/2016:16:56:36.100 +0000] "GET / HTTP/1.1" 200 0 13 "-" "curl/7.30.0" 10.0.2.152:59818 x_forwarded_for:"209.234.137.222" x_forwarded_proto:"http" vcap_request_id:ebc8553c-0b3f-436a-76ee-a81eb1c0248f response_time:0.005669148 app_id:e4437d4f-78ef-4f32-8014-7090d4838371 index: 0
```

Here's what an operator would see tailing the access log on a router VM:
```
dora-test-scoen.superman.cf-app.com - [17/08/2016:17:09:16.135 +0000] "GET / HTTP/1.1" 200 0 13 "-" "curl/7.30.0" 10.0.4.194:50714 x_forwarded_for:"209.234.137.222" x_forwarded_proto:"http" vcap_request_id:320c2001-c753-4b38-6d22-fe7cb10ed1d6 response_time:0.005060497 app_id:e4437d4f-78ef-4f32-8014-7090d4838371 index: 0
```

Here's what an operator would see in a syslog aggregator:
```
Aug 17 10:31:49 10.0.80.70 vcap.gorouter:  [job=router_z2 index=0]  dora-test-scoen.superman.cf-app.com - [17/08/2016:17:31:49.755 +0000] "GET / HTTP/1.1" 200 0 13 "-" "curl/7.30.0" 10.0.4.194:51973 x_forwarded_for:"209.234.137.222" x_forwarded_proto:"http" vcap_request_id:f9887657-efca-40f0-729b-7f15ebdfa9fa response_time:0.004939518 app_id:e4437d4f-78ef-4f32-8014-7090d4838371 index: 0
```

The change is still making its way through CI pipelines and will be delivering in an upcoming final release of [diego-release](https://github.com/cloudfoundry/diego-release), as the change was to the route-emitter component that currently resides in that release. I have a question to the team about changing the field label from `index` to `app_index`; we'll see where they land with that. Feedback welcome.

**Comment by altenhof** [comment]
Hi Shannon,

if I may choose, I’ll go for „app_index“, too.

Regards

Michael


Von: Shannon Coen <notifications@github.com<mailto:notifications@github.com>>
Antworten an: cloudfoundry/gorouter <reply@reply.github.com<mailto:reply@reply.github.com>>
Datum: Mittwoch, 17. August 2016 um 19:38
An: cloudfoundry/gorouter <gorouter@noreply.github.com<mailto:gorouter@noreply.github.com>>
Cc: "Altenhofen, Michael" <michael.altenhofen@sap.com<mailto:michael.altenhofen@sap.com>>, Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Betreff: Re: [cloudfoundry/gorouter] Application Instance in access logs? (#136)


Hello Michael,

Here's what a developer would see in response to cf logs dora:

2016-08-17T09:56:36.10-0700 [RTR/0]      OUT dora-test-scoen.superman.cf-app.com - [17/08/2016:16:56:36.100 +0000] "GET / HTTP/1.1" 200 0 13 "-" "curl/7.30.0" 10.0.2.152:59818 x_forwarded_for:"209.234.137.222" x_forwarded_proto:"http" vcap_request_id:ebc8553c-0b3f-436a-76ee-a81eb1c0248f response_time:0.005669148 app_id:e4437d4f-78ef-4f32-8014-7090d4838371 index: 0


Here's what an operator would see tailing the access log on a router VM:

dora-test-scoen.superman.cf-app.com - [17/08/2016:17:09:16.135 +0000] "GET / HTTP/1.1" 200 0 13 "-" "curl/7.30.0" 10.0.4.194:50714 x_forwarded_for:"209.234.137.222" x_forwarded_proto:"http" vcap_request_id:320c2001-c753-4b38-6d22-fe7cb10ed1d6 response_time:0.005060497 app_id:e4437d4f-78ef-4f32-8014-7090d4838371 index: 0


Here's what an operator would see in a syslog aggregator:

Aug 17 10:31:49 10.0.80.70 vcap.gorouter:  [job=router_z2 index=0]  dora-test-scoen.superman.cf-app.com - [17/08/2016:17:31:49.755 +0000] "GET / HTTP/1.1" 200 0 13 "-" "curl/7.30.0" 10.0.4.194:51973 x_forwarded_for:"209.234.137.222" x_forwarded_proto:"http" vcap_request_id:f9887657-efca-40f0-729b-7f15ebdfa9fa response_time:0.004939518 app_id:e4437d4f-78ef-4f32-8014-7090d4838371 index: 0


The change is still making its way through CI pipelines and will be delivering in an upcoming final release of diego-release<https://github.com/cloudfoundry/diego-release>, as the change was to the route-emitter component that currently resides in that release. I have a question to the team about changing the field label from index to app_index; we'll see where they land with that. Feedback welcome.

**Comment by shashwathi** [comment]
Closing this issue as the corresponding [story](https://www.pivotaltracker.com/story/show/122113849) is accepted.


Regards
Shash
